### PR TITLE
Allow no allocation model

### LINKED ISSUE #1133: Allocation is triggered in non-allocation run
I think. If you run the ribasim-nl model, you get the following error:

```julia
Ribasim.jl on  v2024.2.0~7 [$?] via ஃ 1.10.1 took 2m 21.8s ➜ ./build/ribasim_cli/bin/ribasim nl/hws.toml
┌ Info: Starting a Ribasim simulation.
│   ribasim_version = v"2024.2.0"
│   starttime = 2020-01-01T00:00:00
└   endtime = 2021-01-01T00:00:00
ERROR: Invalid allocation network ID 0.
Stacktrace:
  [1] error(s::String)
    @ Base ~/code/Ribasim.jl/build/ribasim_cli/lib/julia/sys.dylib:-1
  [2] get_allocation_model(p::Ribasim.Parameters{PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, drainage::Vector{Union{Missing, Float64}}, potential_evaporation::Vector{Union{Missing, Float64}}, infiltration::Vector{Union{Missing, Float64}}, precipitation::Vector{Union{Missing, Float64}}, urban_runoff::Vector{Union{Missing, Float64}}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, level::Vector{Float64}, flow_rate::Vector{Float64}}}, allocation_network_id::Int64)
    @ Ribasim ~/code/Ribasim.jl/core/src/callback.jl:310
  [3] set_fractional_flow_in_allocation!(p::Ribasim.Parameters{PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, drainage::Vector{Union{Missing, Float64}}, potential_evaporation::Vector{Union{Missing, Float64}}, infiltration::Vector{Union{Missing, Float64}}, precipitation::Vector{Union{Missing, Float64}}, urban_runoff::Vector{Union{Missing, Float64}}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, level::Vector{Float64}, flow_rate::Vector{Float64}}}, node_id::Ribasim.NodeID, fraction::Float64)
    @ Ribasim ~/code/Ribasim.jl/core/src/callback.jl:343
  [4] set_control_params!(p::Ribasim.Parameters{PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, drainage::Vector{Union{Missing, Float64}}, potential_evaporation::Vector{Union{Missing, Float64}}, infiltration::Vector{Union{Missing, Float64}}, precipitation::Vector{Union{Missing, Float64}}, urban_runoff::Vector{Union{Missing, Float64}}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, level::Vector{Float64}, flow_rate::Vector{Float64}}}, node_id::Ribasim.NodeID, control_state::String)
    @ Ribasim ~/code/Ribasim.jl/core/src/callback.jl:377
  [5] discrete_control_affect!(integrator::OrdinaryDiffEq.ODEIntegrator{OrdinaryDiffEq.RK4{typeof(OrdinaryDiffEq.trivial_limiter!), typeof(OrdinaryDiffEq.trivial_limiter!), Static.False}, true, ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(storage = 1:94, integral = 95:94)}}}, Nothing, Float64, Ribasim.Parameters{PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, drainage::Vector{Union{Missing, Float64}}, potential_evaporation::Vector{Union{Missing, Float64}}, infiltration::Vector{Union{Missing, Float64}}, precipitation::Vector{Union{Missing, Float64}}, urban_runoff::Vector{Union{Missing, Float64}}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, level::Vector{Float64}, flow_rate::Vector{Float64}}}, Float64, Float64, Float64, Float64, Vector{ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(storage = 1:94, integral = 95:94)}}}}, SciMLBase.ODESolution{Float64, 2, Vector{ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(storage = 1:94, integral = 95:94)}}}}, Nothing, Nothing, Vector{Float64}, Vector{Vector{ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(storage = 1:94, integral = 95:94)}}}}}, SciMLBase.ODEProblem{ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(storage = 1:94, integral = 95:94)}}}, Tuple{Float64, Float64}, true, Ribasim.Parameters{PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, @NamedTuple{node_id::Vector{Int64}, time::Vector{Dates.DateTime}, drainage::Vector{Union{Missing, Float64}}, potential_evaporation::
```

Essentially, in a control run with a fractionalflow present, `set_fractional_flow_in_allocation!` is always run, even if the model doesn't support allocation (all ids missing, so becoming 0 in reading).

### PR DESCRIPTION
**Pull request by visr**
Fixes #1133

### CODE REVIEW

File: core/src/callback.jl [hunk 1]
```diff
@@ -300,14 +300,17 @@ function discrete_control_affect!(
...
+function get_allocation_model(
+    p::Parameters,
+    allocation_network_id::Int,
+)::Union{AllocationModel, Nothing}
     (; allocation) = p
     (; allocation_network_ids, allocation_models) = allocation
     idx = findsorted(allocation_network_ids, allocation_network_id)
-    if isnothing(idx)
-        error("Invalid allocation network ID $allocation_network_id.")
+    return if isnothing(idx)
```
Comment (Hofer-Julian) [inline core/src/callback.jl:310]:
```suggestion
    if isnothing(idx)
```

**Review by SouthEndMusic — COMMENTED** [review]
I (and @evetion) don't like the way this problem is resolved. I much prefer that in `set_control_params!` the function `set_fractional_flow_in_allocation!` is only called if `config.allocation.use_allocation`.

**Comment by visr** [comment]
I though about using the `config` at first, but went for this since we don't have the Config available during runtime, it is not part of the Parameters. And the line `if isnothing(idx)` suggested to me that it was meant to work like this in the first place.

The Config is in the Model struct. I feel like generally it is a nice split to not require it during runtime, and I don't know if it would go along well with autodiff. So I feel like your suggestion would need a large refactor that perhaps should be a separate PR?

Or perhaps we could add a field `active::Bool` to `struct Allocation` so this information ends up in the parameters? Or a function `is_active(allocation::Allocation) = isempty(allocation.allocation_models)`? That could be done in this PR. What do you think?

**Comment by SouthEndMusic** [comment]
Such an inspirational branch name btw ^^

**Comment by visr** [comment]
You ain't seen `nothing` yet.

File: core/src/parameter.jl [hunk 2]
```diff
@@ -85,6 +85,8 @@ struct Allocation
     }
 end
 
+is_active(allocation::Allocation) = isempty(allocation.allocation_models)
```
Comment (SouthEndMusic) [inline core/src/parameter.jl:88]:
```suggestion
is_active(allocation::Allocation) = !isempty(allocation.allocation_models)
```

**Review by SouthEndMusic — CHANGES_REQUESTED** [review]
It should be the opposite right? Seems like we don't test this  yet

File: core/src/parameter.jl (hunk 2 above)
Comment (visr) [reply core/src/parameter.jl]: Ha, my bad.

**Review by SouthEndMusic — APPROVED** [review]

**Pull request merged**